### PR TITLE
[pro#404] Limit number of batches per month

### PR DESCRIPTION
**Pull request by lizconlan**
## Highlights 

Introduces a new config setting of `PRO_MONTHLY_BATCH_LIMIT` for a global default number of batches per month (defaults to 1) as a handy place to change it if we need to. I've also added a new column to the `pro_accounts` table so that we can set different limits for individual users if required. Only allows users (with the correct feature flag set) to create new batch request if they have a positive credit balance, including limitations on promoting draft batches.

The new dashboard section looks like this (in Safari) when batches are available:

screen shot 2018-02-22 at 18 10 16

And like this when you're waiting for the month to roll over (the wording may not be quite right):

screen shot 2018-02-22 at 18 02 43

When you're no longer allowed to send batches, the "Preview and send" button disappears from the draft batch page to prevent new request from being sent.

batch-request-2

In case anyone figures out a way to circumvent that, both `preview` and `create` check batch requests are currently allowed and redirects back to the edit screen with a flash error to explain what happened. Not pretty but hopefully will rarely - if ever - be seen.

screen shot 2018-02-23 at 15 16 00

Similarly, attempts to access the URL for the start of the Pro batch process manually results in being redirected back to the new request form (currently without explanation so this may not be quite right).

## Contested implementation detail - "per calendar month"

I've implemented "per month" based on my understanding of "per calendar month" which has a specific legal meaning in regards to contracts and recurring payments: "where one month means on the same (numbered) day of each month (or a period from e.g. the 15th of one month to the 14th of the next)" [1] (as opposed to defining it to mean a set number of days); generally the account/first payment anniversary date [2]. This allows us to tie usage to a nominal billing cycle that should closely match customers making mostly payments but without relying on the payments feature being enabled, or calls out to Stripe. However, it does add a bunch of messy code to do unpleasant date-based arithmetic.

If this isn't clear enough to customers (or the date-based arithmetic is too painful to keep), the alternative would be "per named month" (or just "each month", which is suitably vague) and, presumably, changing the phrasing wherever we display "You have 2 batches left..." to something like "You have 2 batches remaining for **February**" and the simplified "...refreshes on 1st March" rather than a number of days countdown to the new set of batch requests. (Aside: We should probably be clear somewhere that these benefits don't roll over/accumulate.) The code adjustments should be fairly simple - remove the date faff private methods, make `batch_period_start` return the first day of the current month, make `batch_period_renews` the first day of the following month and ditch `days_to_batch` altogether.

Connects to mysociety/alaveteli-professional#404 (doesn't cover all of the UI changes as some of these are being worked on elsewhere so will need to be picked up afterwards. ToDo: ticket those things)

--- 

[1] http://www.duhaime.org/LegalDictionary/C/CalendarMonth.aspx
[2] Although having had to write something nasty to deal with the 31st February problem, I have some sympathy for the mid-month standardised billing cycle people

### CODE REVIEW

File: db/migrate/20180202150837_add_monthly_batch_limit_to_pro_account.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddMonthlyBatchLimitToProAccount < ActiveRecord::Migration
```
[the remark this answers is not in the document]
Comment (lizconlan) [reply db/migrate/20180202150837_add_monthly_batch_limit_to_pro_account.rb]: ugh, maybe we should add an exception for migrations

[the remark this answers is not in the document]
Comment (garethrees) [reply db/migrate/20180202150837_add_monthly_batch_limit_to_pro_account.rb]: Feel free to add as a commit in this PR if its possible

File: app/models/pro_account.rb [hunk 2]
```diff
@@ -63,4 +81,17 @@ def set_stripe_customer_id
   def stripe_customer!
     Stripe::Customer.retrieve(stripe_customer_id) if stripe_customer_id
   end
+
+  def day_given_month(day, month = Time.zone.now.month)
```
Comment (lizconlan) [inline app/models/pro_account.rb]: naming things is hard

File: app/controllers/alaveteli_pro/batch_request_authority_searches_controller.rb [hunk 3]
```diff
@@ -50,7 +50,9 @@ def check_page_limit!(page, per_page)
   end
 
   def check_user_has_batch_access
-    unless feature_enabled? :pro_batch_access, current_user
+    unless (feature_enabled?(:pro_batch_access, current_user) &&
+            current_user.pro_account &&
+            current_user.pro_account.batches_remaining > 0)
```
[the remark this answers is not in the document]
Comment (lizconlan) [reply app/controllers/alaveteli_pro/batch_request_authority_searches_controller.rb]: oh `nonzero?` isn't truthy, that's irritating (returns nil or a number)

[the remark this answers is not in the document]
Comment (garethrees) [reply app/controllers/alaveteli_pro/batch_request_authority_searches_controller.rb]: Yeah, see https://bugs.ruby-lang.org/issues/9123

**Comment by garethrees** [comment]
Superseded by https://github.com/mysociety/alaveteli/pull/4659

**Pull request closed without merging**